### PR TITLE
Use `available_parallelism` instead of `num_cpus`

### PR DESCRIPTION
**Pull request by weihanglo**
**Status**: blocked on https://github.com/rust-lang/rust/pull/92697

---

Just realized that we can drop one dependency 🎉 

Also, we don't want to block timing data output, so if parallelism data is not available, the table will display `ncpu=n/a` instead.

image

### CODE REVIEW

File: src/cargo/core/compiler/build_config.rs [hunk 1]
```diff
@@ -70,7 +71,10 @@ impl BuildConfig {
                  its environment, ignoring the `-j` parameter",
             )?;
         }
-        let jobs = jobs.or(cfg.jobs).unwrap_or(::num_cpus::get() as u32);
+        let jobs = match jobs.or(cfg.jobs) {
+            Some(j) => j,
+            None => available_parallelism()?.get() as u32,
```
Comment (alexcrichton) [inline src/cargo/core/compiler/build_config.rs:76]: Could the error-propagation here have some context indicating what the error is? (e.g. something like "failed to determine the amount of parallelism available")

**Review by alexcrichton — APPROVED** [review]

**Comment by weihanglo** [comment]
Thanks for the review. I think @ehuss is right. `num_cpus` takes cgroups into consideration but `available_parallelism` isn't. Closing.

Ref:

- https://github.com/seanmonstar/num_cpus/blob/5f1b03332000b4c4274b5bd35fac516049ff1c6b/src/linux.rs#L33-L36
- https://github.com/rust-lang/rust/blob/1.59.0/library/std/src/sys/unix/thread.rs#L280-L292

**Comment by joshtriplett** [comment]
@ehuss @weihanglo `available_parallelism` currently has a PR under review to take cgroups into account. I'd suggest leaving this PR open and blocking it on https://github.com/rust-lang/rust/pull/92697 .

**Comment by joshtriplett** [comment]
`available_parallelism` in current std now respects cgroups, so we can switch to it without any regressions.

@bors r+

**Pull request merged**